source/core/rtc_adapter/VideoReceiveAdapter.cc

```diff
@@ -280,11 +278,43 @@ std::unique_ptr<webrtc::VideoDecoder> VideoReceiveAdapterImpl::CreateVideoDecode
 
 int VideoReceiveAdapterImpl::onRtpData(char* data, int len)
 {
-    std::shared_ptr<DataPacket> wp = std::make_shared<DataPacket>(data, len);
-    taskQueue()->PostTask([this, wp]() {
+    rtc::CopyOnWriteBuffer buffer(data, len);
+    if (m_format == FRAME_FORMAT_VP9 &&
+        (m_preferredTemporalId >= 0 || m_preferredSpatialId >= 0)) {
+        webrtc::RtpPacket rtpPacket;
+        webrtc::RTPVideoHeader video_header;
+        if (rtpPacket.Parse((const uint8_t*) data, len)) {
+            if (webrtc::VideoRtpDepacketizerVp9::ParseRtpPayload(
+                rtpPacket.PayloadBuffer(), &video_header) > 0) {
+                webrtc::RTPVideoHeaderVP9* vp9_header =
+                    absl::get_if<webrtc::RTPVideoHeaderVP9>(
+                        &(video_header.video_type_header));
+                if (vp9_header) {
+                    if ((m_preferredSpatialId >= 0 &&
+                        vp9_header->spatial_idx <= kMaxSpatialLayers &&
+                        vp9_header->spatial_idx > m_preferredSpatialId) ||
+                        (m_preferredTemporalId >= 0 &&
+                        vp9_header->temporal_idx <= kMaxTemporalLayers &&
+                        vp9_header->temporal_idx > m_preferredTemporalId)) {
+                        return len;
+                    }
+                    if (vp9_header->spatial_idx == m_preferredSpatialId &&
+                        vp9_header->end_of_frame && !rtpPacket.Marker()) {
+                        // Set Marker
+                        rtpPacket.SetMarker(true);
+                        buffer = rtpPacket.Buffer();
+                    }
+                }
+            } else {
+                // Empty payload, vp9 parse fail
```

Comment:
If the parse fails, you should not proceed to deliver packet.

Reply:
RTP packet with empty payload matters